🐾 Process-supervised RM Trainer

gaetanlop opened the pull request:
# What does this PR do?

Adding support for process-supervised reward training to TRL as requested in #2110 .

List of papers using PRMs: [[1]](https://arxiv.org/pdf/2305.20050), [[2]](https://arxiv.org/pdf/2211.14275), [[3]](https://aclanthology.org/2023.acl-long.291.pdf), [[4]](https://arxiv.org/pdf/2308.09583)...

Fixes # (issue)

#2110 

## Before submitting
- [X] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [X] Did you write any new necessary tests?


## Who can review?

@lewtun @kashif

gaetanlop commented:
Hey @lewtun, thank you for the message. Currently, the only files that are more or less ready are `prm_trainer.py` and `prm_config.py`. The rest are just placeholders that I haven’t had the opportunity to work on yet.

Implementing a PRMs seems to be pretty straighforward, it seems to be a token classification task where only prediction for the last token of each step gets assigned a label and other tokens are ignored during loss calculation. 

If the dataset isn’t pre-tokenized, I assume it should contain the following columns:
- **prompt**: Either a string or past messages
- **steps**: A list of strings
- **labels**: A list of integers corresponding to the label associated to each step

Are you aware of an HF dataset to train PRMs for the example file? Also, how can I add a new subset to the `trl-internal-testing/zen` dataset to support stepwise reward models for the unit test of the `prm_trainer`?

Thanks again for your time!

gaetanlop commented:
PR ready for review. 

**Tests:** I created a dummy_dataset but we should add a subset to trl-internal-testing/zen as done in other scripts.
**Example:** The example is currently using a placeholder for the dataset name as to the best of my knowledge trl didn't release a dataset for stepwise reasoning on HF. We should add this too.

lewtun reviewed (commented):
Thank you for the very clean PR @gaetanlop - this looks great! I've left some minor suggestions regarding the structure, but aside from that and having a smallish dataset in the right format we can sanity check that the accuracy goes up, loss goes down etc I think this is quite close to being ready

gaetanlop commented on trl/trainer/stepwise_reward_trainer.py:
```diff
@@ -0,0 +1,255 @@
+
+    for prompt, steps, labels in zip(batch["prompt"], batch["stepwise_completion"], batch["stepwise_labels"]):
+        if len(steps) != len(labels):
+            raise NotImplementedError("StepwiseRewardTrainer does not support training with unlabeled steps.")
```
Indeed, a `ValueError` makes more sense here, since I am also not aware of any cases where training on unlabeled steps would make sense. In all cases, if the user wants to train with an unlabeled step, he could just set the label to -100.

gaetanlop commented on examples/scripts/stepwise_reward_modeling.py:
```diff
@@ -0,0 +1,130 @@
+Full training:
+python examples/scripts/stepwise_reward_modeling.py \
+    --model_name_or_path Qwen/Qwen2-0.5B-Instruct \
+    --dataset_name trl-lib/PLACEHOLDER \
```
I made two pull requests to trl-lib/zen (https://huggingface.co/datasets/trl-lib/zen/discussions/3) to add the subsets to trl-lib.

qgallouedec replied:
https://github.com/huggingface/trl/pull/2127/#issuecomment-2385061422

gaetanlop commented on docs/source/stepwise_reward_trainer.mdx:
```diff
@@ -0,0 +1,56 @@
+# Stepwise Reward Modeling
```
Not the time to do it today, I will reformat soon :)

qgallouedec replied:
@gaetanlop I can help with this, feel free to let me know if you want me to do that

gaetanlop commented:
Thanks for looking at this @lewtun. Seems like trl-internal-testing/zen is the dataset you are using for testing. I have done a PR to trl-lib/zen, should I also PR trl-internal-testing/zen to add 19 samples of PRM800K for testing or are you handling it on your side?

qgallouedec commented on docs/source/_toctree.yml:
```diff
@@ -46,6 +46,8 @@
     - local: reward_trainer
       title: Reward
+    - local: stepwise_reward_trainer
+      title: StepwiseReward
```
```suggestion
      title: Stepwise Reward
```

qgallouedec commented:
> Seems like `trl-internal-testing/zen` is the dataset you are using for testing. I have done a PR to `trl-lib/zen`, should I also PR `trl-internal-testing/zen` to add 19 samples of PRM800K for testing or are you handling it on your side (it looks like they are both the same dataset)?

`trl-lib/zen` is a mistake. I'll remove it. Can you make the PR to `trl-internal-testing/zen`?

qgallouedec commented on docs/source/dataset_formats.mdx:
```diff
@@ -76,6 +76,19 @@ The *format* of a dataset refers to how the data is structured, typically catego
+    </td>
+    <td>
+      <pre><code>{"prompt": [{"role": "system", "content": "You are a very skilled mathematician."}, {"role": "user", "content": "Two apples and one orange cost 1.5 euros. Four apples and one orange cost 2.5 euros. What's the price of an apple?"}],
+ "stepwise_completion": ["Let a represent the price of an apple.", "Let b represent the price of an orange"],
```
```suggestion
 "completion": ["Let a represent the price of an apple.", "Let b represent the price of an orange"],
```

or maybe 

```suggestion
 "completions": ["Let a represent the price of an apple.", "Let b represent the price of an orange"],
```
?

gaetanlop replied:
Asjusted given #2148

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/trl/pr_2127). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

qgallouedec commented on docs/source/dataset_formats.mdx:
```diff
@@ -76,6 +76,19 @@ The *format* of a dataset refers to how the data is structured, typically catego
+    <td>
+      <pre><code>{"prompt": [{"role": "system", "content": "You are a very skilled mathematician."}, {"role": "user", "content": "Two apples and one orange cost 1.5 euros. Four apples and one orange cost 2.5 euros. What's the price of an apple?"}],
+ "stepwise_completion": ["Let a represent the price of an apple.", "Let b represent the price of an orange"],
+ "stepwise_labels": [True, True]}</code></pre>
```
same here

```suggestion
 "label": [True, True]}</code></pre>
```
or 

```suggestion
 "labels": [True, True]}</code></pre>
```

qgallouedec commented on docs/source/dataset_formats.mdx:
```diff
@@ -188,24 +201,33 @@ An unpaired preference dataset is similar to a preference dataset but instead of
+| [`ORPOTrainer`]           | Preference (explicit prompt) |
+| [`PPOv2Trainer`]          | Tokenized language modeling  |
+| [`RewardTrainer`]         | Preference (implicit prompt) |
+| [`StepwiseRewardTrainer`] | Stepwise preference             |
+| [`SFTTrainer`]            | Language modeling            |
```
```suggestion
| [`SFTTrainer`]            | Language modeling            |
| [`StepwiseRewardTrainer`] | Stepwise preference          |
```

qgallouedec commented on examples/scripts/stepwise_reward_modeling.py:
```diff
@@ -0,0 +1,130 @@
+    ##############
+    # Load dataset
+    ##############
+    dataset = load_dataset(args.dataset_name, data_dir="conversational_stepwise_preference")
```
```suggestion
    dataset = load_dataset(args.dataset_name)
```

qgallouedec commented on examples/scripts/stepwise_reward_modeling.py:
```diff
@@ -0,0 +1,130 @@
+
+    # If post-training a base model, use ChatML as the default template
+    if tokenizer.chat_template is None:
+        model, tokenizer = setup_chat_format(model, tokenizer)
```
Note for maintainers: we should use this util in every example (instead of having, for example), just this:

https://github.com/huggingface/trl/blob/de3876577c1cc37e5c16d0982ecb417c4bd6a752/examples/scripts/cpo.py#L93-L94

qgallouedec commented on examples/scripts/stepwise_reward_modeling.py:
```diff
@@ -0,0 +1,130 @@
+
+if __name__ == "__main__":
+    parser = HfArgumentParser((RewardScriptArguments, StepwiseRewardConfig, ModelConfig))
+    args, training_args, model_config = parser.parse_args_into_dataclasses()
```
```suggestion
    script_args, training_args, model_config = parser.parse_args_into_dataclasses()
```

Can you reflect this change throughout the script?

gaetanlop replied:
Done

qgallouedec commented on examples/scripts/stepwise_reward_modeling.py:
```diff
@@ -0,0 +1,130 @@
+    tokenizer = AutoTokenizer.from_pretrained(
+        model_config.model_name_or_path, trust_remote_code=model_config.trust_remote_code, use_fast=True
+    )
+    model = AutoModelForTokenClassification.from_pretrained(
+        model_config.model_name_or_path, num_labels=2, trust_remote_code=model_config.trust_remote_code, **model_kwargs
+    )
```
Why this instead of what we use in the reward trainer script?

```python
model = AutoModelForSequenceClassification.from_pretrained(
        model_config.model_name_or_path, num_labels=1, trust_remote_code=model_config.trust_remote_code, **model_kwargs
    )
```

gaetanlop replied:
We need to use `AutoModelForTokenClassification` as PRMs are trained "to predict the correctness of each step after the last token in each step" (https://arxiv.org/pdf/2305.20050) which means that for each sample we have `x` scores to be returned (where `x` is the number of steps in the sample).

qgallouedec commented on tests/test_stepwise_reward_trainer.py:
```diff
@@ -0,0 +1,178 @@
+    def setUp(self):
+        self.model_id = "hf-internal-testing/tiny-random-LlamaForCausalLM"
+        self.tokenizer = AutoTokenizer.from_pretrained(self.model_id)
+        self.tokenizer.chat_template = "{% for message in messages %}{{'<|im_start|>' + message['role'] + '\n' + message['content'] + '<|im_end|>' + '\n'}}{% endfor %}{% if add_generation_prompt %}{{ '<|im_start|>assistant\n' }}{% endif %}"
```
Use `SIMPLE_CHAT_TEMPLATE` instead (from `from trl.trainer.utils`)

qgallouedec commented on tests/test_stepwise_reward_trainer.py:
```diff
@@ -0,0 +1,178 @@
+            )
+            self.assertDictEqual(trainer.train_dataset[:], dummy_dataset[:])
+
+    def test_preprocessing_standard(self):
```
I think this test can be merged with the previous with `    @parameterized.expand([("conversational_stepwise_preference",), ("standard_stepwise_preference",)])`

gaetanlop replied:
Indeed

qgallouedec commented on trl/__init__.py:
```diff
@@ -86,6 +86,8 @@
         "RandomRankJudge",
         "RewardConfig",
         "RewardTrainer",
+        "StepwiseRewardConfig",
```
Can you keep the alphabetical order?

gaetanlop replied:
Done

qgallouedec commented on trl/__init__.py:
```diff
@@ -181,6 +183,8 @@
         RandomRankJudge,
         RewardConfig,
         RewardTrainer,
+        StepwiseRewardConfig,
```
Same here https://github.com/huggingface/trl/pull/2127/files#r1782338942

qgallouedec commented on trl/trainer/__init__.py:
```diff
@@ -58,6 +58,8 @@
     "ppov2_trainer": ["PPOv2Trainer"],
     "reward_config": ["RewardConfig"],
     "reward_trainer": ["RewardTrainer", "compute_accuracy"],
+    "stepwise_reward_config": ["StepwiseRewardConfig"],
```
same here https://github.com/huggingface/trl/pull/2127/files#r1782338942

qgallouedec commented on trl/trainer/__init__.py:
```diff
@@ -122,6 +124,8 @@
     from .ppov2_trainer import PPOv2Trainer
     from .reward_config import RewardConfig
     from .reward_trainer import RewardTrainer, compute_accuracy
+    from .stepwise_reward_config import StepwiseRewardConfig
```
same here https://github.com/huggingface/trl/pull/2127/files#r1782338942

qgallouedec commented on trl/trainer/stepwise_reward_trainer.py:
```diff
@@ -0,0 +1,297 @@
+                    "max_length": args.max_length,
+                    "step_separator": args.step_separator,
+                }
+                train_dataset = train_dataset.map(maybe_apply_chat_template, fn_kwargs=chat_template_kwargs)
```
add `num_proc=args.dataset_num_proc,` for this one as well. (same for eval dataset)

I realise that it's missing in `RewardTrainer` as well, I'll fix it

qgallouedec commented on trl/trainer/stepwise_reward_trainer.py:
```diff
@@ -0,0 +1,297 @@
+                    eval_dataset = eval_dataset.map(maybe_apply_chat_template, fn_kwargs=chat_template_kwargs)
+                    eval_dataset = eval_dataset.map(
+                        _tokenize,
+                        fn_kwargs=tokenize_kwargs,
+                        batched=True,
```
```suggestion
                        batched=True,
                        fn_kwargs=tokenize_kwargs,

```

qgallouedec commented on trl/trainer/stepwise_reward_trainer.py:
```diff
@@ -0,0 +1,297 @@
+            dataset_name=dataset_name,
+            tags=tags,
+            wandb_url=wandb.run.get_url() if is_wandb_available() and wandb.run is not None else None,
+            trainer_name="StepwiseReward",
```
```suggestion
            trainer_name="Stepwise Reward",
```

qgallouedec commented on trl/trainer/stepwise_reward_trainer.py:
```diff
@@ -0,0 +1,297 @@
+        else:
+            base_model = None
+
+        model_card = generate_model_card(
```
We don't have a paper for this one?

qgallouedec commented on trl/trainer/stepwise_reward_trainer.py:
```diff
@@ -0,0 +1,297 @@
+        model_card.save(os.path.join(self.args.output_dir, "README.md"))
+
+    @wraps(Trainer.push_to_hub)
+    def push_to_hub(
+        self,
+        commit_message: Optional[str] = "End of training",
+        blocking: bool = True,
+        **kwargs,
+    ) -> str:
+        """
+        Overwrite the `push_to_hub` method in order to force-add the tag "stepwise-reward-trainer" when pushing the
+        model on the Hub. Please refer to `~transformers.Trainer.push_to_hub` for more details.
+        Unlike the parent class, we don't use the `token` argument to mitigate security risks.
+        """
+        kwargs = trl_sanitze_kwargs_for_tagging(model=self.model, tag_names=self._tag_names, kwargs=kwargs)
+        return super().push_to_hub(commit_message=commit_message, blocking=blocking, **kwargs)
```
```suggestion
```

qgallouedec commented on trl/trainer/utils.py:
```diff
@@ -743,13 +743,23 @@ def get_global_statistics(
-    if np.array(predictions[:, 0] == predictions[:, 1], dtype=float).sum() > 0:
-        warnings.warn(
-            f"There are {np.array(predictions[:, 0] == predictions[:, 1]).sum()} out of {len(predictions[:, 0])} instances where the predictions for both options are equal. As a consequence the accuracy can be misleading."
+    if predictions.ndim == 3:
```
We need to have a test for this new case

gaetanlop replied:
Yes, there is already one here: https://github.com/gaetanlop/trl/blob/c3eb08ea31951a40f1d7cad45e4ff433bc5fa5c2/tests/test_stepwise_reward_trainer.py#L39-L45

Shouldn't it be enough given that the `RewardTrainer` also has a test when ndim==2?

qgallouedec commented on trl/trainer/stepwise_reward_trainer.py:
```diff
@@ -0,0 +1,297 @@
+                    "When the dataset isn't pretokenized, you should set `max_length` in the `StepwiseRewardConfig`"
+                    " we have set it for you, but you should do it yourself in the future."
+                )
+            if args.step_separator is None:
+                args.step_separator = "\n"
+                warnings.warn(
+                    "When the dataset isn't pretokenized, you should set `step_separator` in the `StepwiseRewardConfig`"
+                    " we have set it for you to '\n', but you should do it yourself in the future."
+                )
```
```suggestion
```

It can't be `None`, see the config

qgallouedec commented on trl/trainer/stepwise_reward_config.py:
```diff
@@ -0,0 +1,41 @@
+    """
+
+    max_length: Optional[int] = None
+    step_separator: Optional[str] = None
```
```suggestion
    step_separator: str = "\n"
```

qgallouedec commented on trl/trainer/stepwise_reward_config.py:
```diff
@@ -0,0 +1,41 @@
+    Parameters:
+        max_length (`int`, *optional*, defaults to `None`):
+            Maximum length of the sequences (prompt + steps) in the batch.
+        post_step_separator (`str`, *optional*, defaults to `\n`):
```
```suggestion
        post_step_separator (`str`, *optional*, defaults to `"\n"`):
```

qgallouedec commented on trl/trainer/stepwise_reward_config.py:
```diff
@@ -0,0 +1,41 @@
+@dataclass
+class StepwiseRewardConfig(TrainingArguments):
+    r"""
+    Configuration class for the [`StepwiseRewardConfig`].
```
```suggestion
    Configuration class for the [`StepwiseRewardTrainer`].
```

qgallouedec commented on trl/trainer/stepwise_reward_config.py:
```diff
@@ -0,0 +1,41 @@
+
+    Parameters:
+        max_length (`int`, *optional*, defaults to `None`):
+            Maximum length of the sequences (prompt + steps) in the batch.
```
```suggestion
            Maximum length of the sequences (prompt + completion) in the batch. The completion is the concatenation of the steps.
```

qgallouedec commented on examples/scripts/stepwise_reward_modeling.py:
```diff
@@ -0,0 +1,130 @@
+Full training:
+python examples/scripts/stepwise_reward_modeling.py \
+    --model_name_or_path Qwen/Qwen2-0.5B-Instruct \
+    --dataset_name trl-lib/zen \
```
use another dataset, this one intended for unit testing

gaetanlop replied:
Done

qgallouedec commented on examples/scripts/stepwise_reward_modeling.py:
```diff
@@ -0,0 +1,130 @@
+LoRA:
+python examples/scripts/stepwise_reward_modeling.py \
+    --model_name_or_path Qwen/Qwen2-0.5B-Instruct \
+    --dataset_name trl-lib/zen \
```
same here https://github.com/huggingface/trl/pull/2127/files#r1782398655

qgallouedec commented:
Thanks a lot @gaetanlop! I left some comments.
I think we need to first define a new standard for the dataset required here. I prefer working on it an another PR: #2148.
Once we've merged #2148 we'll need to update this one.

Feel free to share your thoughts in #2148 btw.

gaetanlop commented:
Thanks for the reviews @qgallouedec. I will wait that #2148 is completed and merged before resolving your comments. Don't hesitate to ping me here when you have done your decisions regarding the new standard dataset format.

gaetanlop commented on docs/source/dataset_formats.mdx:
```diff
@@ -76,6 +76,19 @@ The *format* of a dataset refers to how the data is structured, typically catego
+  <tr>
+    <td>Stepwise preference</td>
+    <td>
+      <pre><code>{"prompt": "Two apples and one orange cost 1.5 euros. Four apples and one orange cost 2.5 euros. What's the price of an apple?",
```
Yes of course, it will makes less sense though

yiyepiaoling0715 commented:
hi,a good job?  when  will this be merged?

qgallouedec commented on docs/source/dataset_formats.mdx:
````diff
@@ -250,25 +250,34 @@ stepwise_example = {
 }
 ```
 
+### Stepwise preference
+
+A stepwise preference dataset is similar to an unpaired preference dataset but instead of having a single `"completion"` and `"label"`, it includes a `"completion"` column that splits the completion into a list of steps and a `"labels"` column indicating whether each step is correct or not.
+
+```python
+steps_preference_example = {"prompt": "The sky is", "completion": [", let me think...", "blue."], "labels": [False, True]}
+```
+
````
```suggestion
```

Remove in favour of "Stepwise supervision"

qgallouedec commented on examples/datasets/openai_prm800k.py:
```diff
@@ -0,0 +1,54 @@
+# Copyright 2024 The HuggingFace Inc. team. All rights reserved.
```
You can remove this file in favour of https://github.com/huggingface/trl/blob/main/examples/datasets/prm800k.py

qgallouedec commented on trl/trainer/stepwise_reward_trainer.py:
```diff
@@ -0,0 +1,298 @@
+            peft_config (`Dict`, defaults to `None`):
+                The PEFT configuration to use for training. If you pass a PEFT configuration, the model will be wrapped in a PEFT model.
+        """
+        if type(args) is not StepwiseRewardConfig:
+            raise ValueError(f"args should be an instance of `StepwiseRewardConfig` but got {type(args)}")
```
```suggestion
```

qgallouedec commented on trl/trainer/stepwise_reward_trainer.py:
```diff
@@ -0,0 +1,298 @@
+
+        citation = textwrap.dedent(
+            """\
+        @article{uesato2022solving,
+        title={Solving math word problems with process-and outcome-based feedback},
+        author={Uesato, Jonathan and Kushman, Nate and Kumar, Ramana and Song, Francis and Siegel, Noah and Wang, Lisa and Creswell, Antonia and Irving, Geoffrey and Higgins, Irina},
+        journal={arXiv preprint arXiv:2211.14275},
+        year={2022}
+        }"""
```
```suggestion
        @article{uesato2022solving,
	title        = {Solving Math Word Problems With Process- and Outcome-Based Feedback},
	author       = {Uesato, Jonathan and Kushman, Nate and Kumar, Ramana and Song, Francis and Siegel, Noah and Wang, Lisa and Creswell, Antonia and Irving, Geoffrey and Higgins, Irina},
	year         = 2022,
	journal      = {arXiv preprint arXiv:2211.14275}
        }"""
```

lewtun commented on docs/source/dataset_formats.mdx:
```diff
@@ -250,25 +250,34 @@ stepwise_example = {
+| [`PPOTrainer`]            | Tokenized language modeling                                                                            |
+| [`RewardTrainer`]         | [Preference (implicit prompt recommended)](#preference)                                                |
+| [`SFTTrainer`]            | [Language modeling](#language-modeling)                                                                |
+| [`StepwiseRewardTrainer`] | [Stepwise preference](#stepwise-preference)                                                            |
```
```suggestion
| [`StepwiseRewardTrainer`] | [Stepwise supervision](#stepwise-supervision)                                                            |
```

lewtun commented on docs/source/stepwise_reward_trainer.mdx:
```diff
@@ -0,0 +1,85 @@
+
+## Usage tips  
+
+The [`StepwiseRewardTrainer`] is a wrapper around the [`Trainer`] class. It needs two parameters to be set via the [`StepwiseRewardConfig`] namely:
```
turbo nit:

```suggestion
The [`StepwiseRewardTrainer`] is a wrapper around the [`Trainer`] class. It needs two parameters to be set via the [`StepwiseRewardConfig`], namely:
```

lewtun commented on docs/source/stepwise_reward_trainer.mdx:
```diff
@@ -0,0 +1,85 @@
+## Usage tips  
+
+The [`StepwiseRewardTrainer`] is a wrapper around the [`Trainer`] class. It needs two parameters to be set via the [`StepwiseRewardConfig`] namely:
+* `max_length`: controls the maximum length of the sequences where a sequence is composed of the prompt and the concatenation of each completion steps.
```
nit:
```suggestion
* `max_length`: controls the maximum length of the sequences, where a sequence is composed of the prompt and the concatenation of each completion step.
```

lewtun commented on docs/source/stepwise_reward_trainer.mdx:
```diff
@@ -0,0 +1,85 @@
+
+## Expected dataset format
+
+The dataset should be formatted as a [Name to find](dataset_formats#[Name to find]) which implies that the dataset should contain the following columns: `prompt`, `completion` and `labels` where `completion` contains a list of reasoning steps and `labels` a list of booleans indicating the correctness of each step.
```
```suggestion
The dataset should be formatted as a [Stepwise Supervision](dataset_formats#stepwise-supervision) dataset, which implies that it should contain the following columns: `prompt`, `completions` and `labels`, where `completions` contains a list of reasoning steps and `labels` a list of booleans or floats indicating the correctness of each step.
```

lewtun commented on examples/scripts/stepwise_reward_modeling.py:
```diff
@@ -0,0 +1,133 @@
+Full training:
+python examples/scripts/stepwise_reward_modeling.py \
+    --model_name_or_path Qwen/Qwen2-0.5B-Instruct \
+    --dataset_name trl-lib/openai-prm800k-15k \
```
```suggestion
    --dataset_name trl-lib/prm800k \
```

lewtun commented on examples/scripts/stepwise_reward_modeling.py:
```diff
@@ -0,0 +1,133 @@
+LoRA:
+python examples/scripts/stepwise_reward_modeling.py \
+    --model_name_or_path Qwen/Qwen2-0.5B-Instruct \
+    --dataset_name trl-lib/openai-prm800k-15k \
```
```suggestion
    --dataset_name trl-lib/prm800k \
```

lewtun commented on examples/scripts/stepwise_reward_modeling.py:
```diff
@@ -0,0 +1,133 @@
+        model_config.model_name_or_path, trust_remote_code=model_config.trust_remote_code, use_fast=True
+    )
+    model = AutoModelForTokenClassification.from_pretrained(
+        model_config.model_name_or_path, num_labels=3, trust_remote_code=model_config.trust_remote_code, **model_kwargs
```
With the new format, shouldn't this be just two labels?
```suggestion
        model_config.model_name_or_path, num_labels=2, trust_remote_code=model_config.trust_remote_code, **model_kwargs
```

gaetanlop replied:
Indeed

lewtun commented on examples/scripts/stepwise_reward_modeling.py:
```diff
@@ -0,0 +1,133 @@
+    --max_length 2048
+
+LoRA:
+python examples/scripts/stepwise_reward_modeling.py \
```
If you have some compute, can you share some WandB logs from running these scripts? Otherwise I can run them myself :)

lewtun reviewed (commented):
Thanks for iterating @gaetanlop and apologies for the slow review on this one 🙈 ! Overall this is looking really good and with some minor changes I think it's close to being ready

gaetanlop commented:
Hey @lewtun @qgallouedec, thanks for the reviews. I will incorporate your comments by the end of the week. I should be able to run the experiments on my side and share some logs here.

gaetanlop commented on docs/source/stepwise_reward_trainer.mdx:
```diff
@@ -0,0 +1,85 @@
+
+The [`StepwiseRewardTrainer`] is a wrapper around the [`Trainer`] class. It needs two parameters to be set via the [`StepwiseRewardConfig`] namely:
+* `max_length`: controls the maximum length of the sequences where a sequence is composed of the prompt and the concatenation of each completion steps.
+* `step_separator`: indicate the separator used to separate each step of the reasoning process. By default, it is set to `"n"`.
```
Yes it should, this was a typo, thanks for spotting it

gaetanlop commented on docs/source/stepwise_reward_trainer.mdx:
```diff
@@ -0,0 +1,85 @@
+
+model = AutoModelForTokenClassification.from_pretrained("Qwen/Qwen2-0.5B-Instruct", num_labels=2)
+
+train_dataset = Dataset.from_dict(
```
Yes I agree

gaetanlop commented:
1. I think the user should be able to train on **the last step only** (still framing it as a token classification task). 
Looking at [prm800k](https://huggingface.co/datasets/trl-lib/prm800k), training the model to predict the correctness of each step in a solution means that we will train the very same prefix solution multiple times (look at the first 6 rows for example). A better approach for this dataset seems to be training the model to predict the correctness of only the last step in the solution, not the intermediate ones. For this approach, if the model is autoregressive, scoring all steps of a solution in one forward pass during inference is still possible without training-inference mismatch.
The sentence from the OpenAI paper [Let's verify step by step](https://arxiv.org/pdf/2305.20050) seems to support this: “We train our PRMs by fine-tuning the MathMix model to predict the probability of positive, negative, and neutral labels **given a solution prefix ending in one of our labeled steps.**” (refer to Appendix F1 for more details). 

This would require either letting the user specify an ignore step value in labels (1, 0 or -100) or just specify that we want to train on the last step only in the config. What's your preference?

2. PR #2148 do not introduce any conversational dataset format for Stepwise trainings. Is this an oversight or is this on purpose? I think supporting conversational datasets is important as reward models are usually initialized with the sft models.

3. The dataset config `standard_step` for `zen` seems to not be available. The tests won't pass. I have let the PR in DRAFT mode.

qgallouedec commented on docs/source/stepwise_reward_trainer.mdx:
```diff
@@ -0,0 +1,85 @@
+
+The dataset should be formatted as a [Stepwise Supervision](dataset_formats#stepwise-supervision) dataset, which implies that it should contain the following columns: `prompt`, `completions` and `labels`, where `completions` contains a list of reasoning steps and `labels` a list of booleans or floats indicating the correctness of each step.
+
+The [`StepwiseRewardTrainer`] only supports [standard](dataset_formats#standard) dataset format. When provided with a conversational dataset, the trainer will automatically apply the chat template to the dataset.
```
```suggestion
The [`StepwiseRewardTrainer`] only supports [standard](dataset_formats#standard) dataset format.
```

qgallouedec commented:
We now have have a dataset for testing:

```python
>>> from datasets import load_dataset
>>> dataset = load_dataset("trl-internal-testing/zen", "standard_stepwise_supervision")
>>> dataset["train"][4]
{'prompt': 'Flat is better than', 'completions': [' in terms of structure,', ' nested.'], 'label': [True, False]}
```

gaetanlop commented:
@qgallouedec the term used for the `labels` in standard_stepwise_supervision and prm800K differs, prm800K uses `labels` while standard_stepwise_supervision uses `label`.

qgallouedec commented:
@gaetanlop about the tokenize function, it would be better to 
- have it as a static method 
- using it with a single example (not a batch). 

In fact take example from the implementation of `DPOTrainer`. I can fix this quickly. I offer to do a PR on your branch, wdyt?

gaetanlop commented:
@qgallouedec yes of course, as you prefer, I was following the implementation done in the `KTOTrainer`. You can open a PR otherwise I will make the changes later today

qgallouedec commented:
Should we add the separator token between the prompt and the first step? If you don't (like the current code) you get something like:

```python
prompt = "This is my prompt."
completions = ["This is my first step.", "This is my second step."]
separator = "\n"

# Processing here

result == ""This is my prompt.This is my first step. This is my second step."
```

qgallouedec commented:
https://github.com/gaetanlop/trl/pull/1

I still need to add the collator and then it's ready

qgallouedec commented:
First trained model: https://huggingface.co/qgallouedec/Qwen2-0.5B-Reward

<img width="1358" alt="Screenshot 2024-11-26 at 21 51 58" src="https://github.com/user-attachments/assets/9ea5ee75-6d4c-4ac6-b691-8c254081e500">

gaetanlop commented:
Thank you for the refactoring work on the tokenize_row function, @qgallouedec. I have made some adjustments to ensure proper handling of special tokens. Also, I refined the label creation process and updated the tokenize_row function to support truncation based on both max_length and max_completion_ids. I have added some tests to confirm that the updated tokenize_row function behaves as intended.

I believe the PR is ready

kashif commented on docs/source/stepwise_reward_trainer.mdx:
```diff
@@ -0,0 +1,85 @@
+
+## Overview
+
+Stepwise or process reward models were proposed in [Solving math word problems with processand outcome-based feedback](https://arxiv.org/pdf/2211.14275) by Jonathan Uesato, Nate Kushman, Ramana Kumar, Francis Song, Noah Siegel, Lisa Wang, Antonia Creswell, Geoffrey Irving and Irina Higgins.
```
```suggestion
Stepwise or process reward models were proposed in [Solving math word problems with process and outcome-based feedback](https://huggingface.co/papers/2211.14275) by Jonathan Uesato, Nate Kushman, Ramana Kumar, Francis Song, Noah Siegel, Lisa Wang, Antonia Creswell, Geoffrey Irving, and Irina Higgins.
```

qgallouedec commented:
Thanks, good catch aa33e6292a4b65df5c0cfd478722a0bfb6c11fcc! I don't know how the training code worked without it, maybe with the padding?

> to ensure proper handling of special tokens

Yes that was my question with 228aa31268b40a638a59f07bd5578670bb9bd850. Do you have any example of decoder model that have a bos_token?

qgallouedec commented:
With the fix it seems to overfit. Is it related to the redundancy in the data?

![Screenshot 2024-11-27 at 12 55 46](https://github.com/user-attachments/assets/7e812757-a555-42b0-b083-9d30fd034202)

gaetanlop commented:
> Hey it might be good to run a test on the Math-Shepherd dataset: https://huggingface.co/datasets/peiyi9979/Math-Shepherd
> 
> It will need to be formatted in our schema, but has the advantage of not being derived from multi-step human preferences and should give a clearer signal / speed up debugging

Yes good idea, would you like this dataset to be in `trl-lib` ?

qgallouedec commented:
> Yes good idea, would you like this dataset to be in trl-lib ?

I'm on it

qgallouedec commented:
Finally... So data cleaning was needed: https://huggingface.co/datasets/trl-lib/math_shepherd

```python
>>> from datasets import load_dataset
>>> 
>>> ds = load_dataset("trl-lib/math_shepherd")
>>> ds["train"][0]
{'prompt': 'Janet pays $40/hour for 3 hours per week of clarinet lessons and $28/hour for 5 hours a week of piano lessons. How much more does she spend on piano lessons than clarinet lessons in a year?', 'completions': ['Step 1: Janet spends 3 hours + 5 hours = <<3+5=8>>8 hours per week on music lessons.', 'Step 2: She spends 40 * 3 = <<40*3=120>>120 on clarinet lessons per week.', 'Step 3: She spends 28 * 5 = <<28*5=140>>140 on piano lessons per week.', 'Step 4: Janet spends 120 + 140 = <<120+140=260>>260 on music lessons per week.', 'Step 5: She spends 260 * 52 = <<260*52=13520>>13520 on music lessons in a year. The answer is: 13520'], 'labels': [True, True, True, True, False]}
```

qgallouedec commented:
> Isn't it the case for llama 3.2 for example?

Indeed!

```python
>>> from transformers import AutoTokenizer
>>> tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-3.2-1B-Instruct")
>>> tokenizer("Hey")
{'input_ids': [128000, 19182], 'attention_mask': [1, 1]}
>>> tokenizer("Hey", add_special_tokens=False)
{'input_ids': [19182], 'attention_mask': [1]}
>>> tokenizer.decode(128000)
'<|begin_of_text|>'
>>> tokenizer.bos_token
'<|begin_of_text|>'
```

qgallouedec commented:
Very nice curves 
<img width="1101" alt="Screenshot 2024-11-29 at 10 09 56" src="https://github.com/user-attachments/assets/d7ab2722-b272-42c9-adbd-1e9c29009179">

qgallouedec commented:
Some examples:

```python
from datasets import load_dataset
from transformers import pipeline

pipe = pipeline("token-classification", model="qgallouedec/Qwen2-0.5B-Reward-Math-Sheperd")
dataset = load_dataset("trl-lib/math_shepherd")
example = dataset["test"][10]

scores = []
print("\n".join((example["prompt"], *example["completions"])))
for idx in range(1, len(example["completions"])+1):
    text = "\n".join((example["prompt"], *example["completions"][0:idx])) + "\n"
    score = float(pipe(text)[-1]["score"])
    print(f"Step {idx}\tScore: {score:.2f}\tLabel: {example['labels'][idx-1]}")
```

```
Grandma gave Bryce and Carter some raisins. Bryce received 6 more raisins than Carter, and Carter received half the number of raisins Bryce received. How many raisins did Bryce receive?
Step 1: Let $b$ be the number of raisins Bryce received and $c$ be the number of raisins Carter received.
Step 2: We are given that $b = c + 6$ and $c = \frac{1}{2}b$.
Step 3: Substituting the second equation into the first equation, we get $b = c + 6 = \frac{1}{2}b + 6$.
Step 4: Simplifying, we have $b = \frac{1}{2}b + 6$.
Step 5: Subtracting $\frac{1}{2}b$ from both sides, we get $\frac{1}{2}b - b = 6$.
Step 6: Simplifying further, we have $\frac{1}{2}b - 2b = 6$.
Step 7: Combining like terms, we have $-\frac{1}{2}b = 6$.
Step 8: Multiplying both sides by $-2$, we get $b = -12$.
Step 9: Therefore, Bryce received $\boxed{-12}$ raisins.The answer is: -12
Step 1  Score: 1.00     Label: True
Step 2  Score: 1.00     Label: True
Step 3  Score: 1.00     Label: True
Step 4  Score: 0.99     Label: True
Step 5  Score: 0.98     Label: True
Step 6  Score: 0.82     Label: False
Step 7  Score: 0.82     Label: False
Step 8  Score: 0.87     Label: False
Step 9  Score: 0.94     Label: False
```


```
Dylan filled his glass with ice cubes before he poured his soda into it. He counted and there were 8 ice cubes in his glass. Later when he prepares a pitcher of lemonade he puts two times as many ice cubes in the pitcher. Then it is time to refill the ice cube trays which have 12 spaces each for ice cubes. How many trays does Dylan need to fill if he used exactly all the ice they had in them?
Step 1: Dylan used 8 ice cubes in his glass and he used 2 * 8 = <<8*2=16>>16 ice cubes in the pitcher.
Step 2: There are 12 spaces in each ice cube tray and he used 8 + 16 = <<8+16=24>>24 ice cubes.
Step 3: Dylan needs to fill 24 / 12 = <<24/12=2>>2 ice cube trays. The answer is: 2
Step 1  Score: 1.00     Label: True
Step 2  Score: 0.92     Label: True
Step 3  Score: 0.87     Label: True
```

```
Jeremy buys 3 bags of chips for a party. Stacy and Emily also buy chips for the party. If they need 10 bags of chips total, and Stacy buys 4 bags, how many bags of chips should Emily buy?
Step 1: Jeremy and Stacy have already bought 4 + 3 = <<4+3=7>>7 bags of chips.
Step 2: Thus, Emily needs to buy 10 - 7 = <<10-7=3>>3 bags of chips. The answer is: 3
Step 1  Score: 1.00     Label: True
Step 2  Score: 1.00     Label: True
```

```
A charity is delivering chicken and rice dinners to a shelter for the hungry. They have a hundred plates to deliver. The rice cost ten cents a plate and the chicken cost forty cents a plate. How many dollars did the charity spend on the food for the dinners?
Step 1: They spent 10 + 40 = <<10+40=50>>50 cents per plate on food.
Step 2: So they spent 50 x 100 = <<50*100=50>>50 dollars on the food for the dinners. The answer is: 50
Step 1  Score: 0.97     Label: True
Step 2  Score: 0.55     Label: True  # <- here the model right and the label wrong
```

qgallouedec commented:
There was a tricky bug in the processing: the label column was casted to its original dtype (`bool`) after the map, instead of staying in its new dtype (`int`). It implied that `-100` were converted to `True`.

```python
# what we want after tokenization
{"input_ids": [2, 4, 6, ...], "labels": [-100, -100, 0, -100, 1, ...]}
# what we got after tokenization
{"input_ids": [2, 4, 6, ...], "labels": [True, True, False, True, True, ...]}
```

Consequently, instead of being ignored in the loss computation, most tokens were associated to the label `True`.

It explains why we got such high values for accuracy right from start, and why the model predicted most of the time `True` in evaluation.

It has been fixed in 049fdf9.

qgallouedec commented:
New results look way better:

![Screenshot 2024-12-09 at 19 41 30](https://github.com/user-attachments/assets/7993ccb1-c333-44a2-aa03-1425eb7ab874)

qgallouedec commented on docs/source/stepwise_reward_trainer.mdx:
```diff
@@ -0,0 +1,85 @@
+
+> Recent work has shown that asking language models to generate reasoning steps improves performance on many reasoning tasks. When moving beyond prompting, this raises the question of how we should supervise such models: outcome-based approaches which supervise the final result, or process-based approaches which supervise the reasoning process itself? Differences between these approaches might na ...
+
+This post-training method was contributed by [Gaetan Lopez](https://github.com/gaetanlop), [Lewis Tunstall](https://huggingface.co/lewtun) and [Quentin Gallouédec](https://huggingface.co/qgallouedec)
```
```suggestion
This post-training method was contributed by [Gaetan Lopez](https://github.com/gaetanlop), [Lewis Tunstall](https://huggingface.co/lewtun), [Quentin Gallouédec](https://huggingface.co/qgallouedec) and [Agustín Piqueres](https://huggingface.co/plaguss).
```

gaetanlop commented:
The new curves look way more reasonable! Thanks for finding the bug @qgallouedec

skepsun commented:
Thank you for your contribution to the development of the well-written stepwise RM trainer. To further advance RLHF with PRM, some RL trainers, such as the PPO trainer, could potentially benefit from PRMs trained using a stepwise RM trainer. Several points may be considered:

1. A suitable `step_separator` must be defined, ensuring it is always tokenized into a fixed token ID.
2. Implementing a generate function that outputs one step at a time is unnecessary. Instead, an SFT model trained on a CoT dataset (with a defined `step_separator` in the answers) is essential to prepare PPO with PRM. Thus, CoT SFT datasets should be constructed, albeit in small quantities, as SFT is only performed to ensure the model outputs CoT answers in a specific format.
3. The `get_reward` function already supports scoring across all positions, as it is also utilized by the value model. To integrate PRM scores, an additional mask can be used to extract scores at `step_separator` positions and directly add them to the `non_score_reward`.
4. The `score` column in the wandb table should include lists of PRM scores with varying lengths.

gaetanlop commented:
Sounds good to me, that's how I named it in my initial commits.

qgallouedec commented on docs/source/prm_trainer.mdx:
```diff
@@ -0,0 +1,117 @@
+
+## Overview
+
+Process reward models were proposed in [Solving math word problems with process- and outcome-based feedback](https://huggingface.co/papers/2211.14275) by Jonathan Uesato, Nate Kushman, Ramana Kumar, Francis Song, Noah Siegel, Lisa Wang, Antonia Creswell, Geoffrey Irving, and Irina Higgins.
```
Do any of these guys have an HF account?

qgallouedec replied:
Seems like not

qgallouedec commented on trl/trainer/prm_trainer.py:
```diff
@@ -0,0 +1,329 @@
+
+
+class PRMTrainer(Trainer):
+    _tag_names = ["trl", "prm-trainer"]
```
```suggestion
    _tag_names = ["trl", "prm"]
```
